libbpf-tools/map_helpers.c

```diff
@@ -0,0 +1,105 @@
+// SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause)
+// Copyright (c) 2020 Anton Protopopov
+#include <stdlib.h>
+#include <stdio.h>
+#include <errno.h>
+
+#include "map_helpers.h"
+
+#define warn(...) fprintf(stderr, __VA_ARGS__)
+
+static bool batch_map_ops = true; /* hope for the best */
+
+static int
+dump_hash_iter(int map_fd, void *keys, __u32 key_size,
+	       void *values, __u32 value_size, __u32 *count,
+	       void *invalid_key)
+{
+	__u8 key[key_size], next_key[key_size];
+	__u32 n = 0;
+	int err;
+
+	/* First get keys */
+	__builtin_memcpy(key, invalid_key, key_size);
+	while (n < *count) {
+		err = bpf_map_get_next_key(map_fd, key, next_key);
+		if (err && errno != ENOENT) {
+			return -1;
+		} else if (err) {
+			break;
+		}
+		__builtin_memcpy(key, next_key, key_size);
+		__builtin_memcpy(keys + key_size*(n++), next_key, key_size);
+	}
+
+	/* Now read values */
+	for (int i = 0; i < n; i++) {
+		err = bpf_map_lookup_elem(map_fd, keys + key_size*i,
+					  values + value_size*i);
+		if (err)
+			return -1;
+	}
+
+	*count = n;
+	return 0;
+}
+
+static int
+dump_hash_batch(int map_fd, void *keys, __u32 key_size,
+		void *values, __u32 value_size, __u32 *count)
+{
+	void *in = NULL, *out;
+	__u32 n, n_read = 0;
+	int err = 0;
+
+	while (n_read < *count && !err) {
+		n = *count - n_read;
+		err = bpf_map_lookup_batch(map_fd, &in, &out,
+					   keys + n_read*key_size,
+					   values + n_read*value_size,
+					   &n, NULL);
+		if (err && errno != ENOENT) {
+			return -1;
+		}
+		n_read += n;
+		in = out;
```

Comment:
I'm not exactly familiar with map batch API, but aren't you providing the same starting key on each retry, while also advancing keys/values pointers above (with n_read ever increasing). It seem to me like you'll potentially be just getting duplicated keys.

I got an impression that really the only good way to use batching is to assume that batch will return all keys in one go. And if you want to retry, then you need to retry from scratch.

@yonghong-song, do you have an opinion here?

Reply:
Would be very good if someone check this :) However, I hope I am doing this right: `in` and `out` aren't keys, but `batch` ids (or whatever they are called); in case of hashtab they correspond to bucket numbers, and the syscall returns the next bucket to read from in the `batch_out`, see `__htab_map_lookup_and_delete_batch`

Reply:
Yes, `batch` is an opaque number. The `out_batch` is the `in_batch` for the next syscall so the above implementation is correct. The only issue is the buffer size. If the buffer size is less than the number of
elements in any hash table bucket, you will get an error return. Typically this should not happen though
as we have max_entries number of buckets. With batch API, users won't get duplicates as it moves
forward bucket by bucket.

Reply:
ah, I keep missing that it's not the key, but rather bucket index.